File: APISwitcher/Services/SubscriptionService.cs
using System.Net.Http;
using System.Net.Http.Headers;
using System.Net.Http.Json;
using System.Text.Json;
using System.Text.Json.Nodes;
using APISwitcher.Models;

namespace APISwitcher.Services;

/// <summary>
/// 订阅查询服务
/// </summary>
public class SubscriptionService : IDisposable
{
    private readonly HttpClient _httpClient;

    public SubscriptionService()
    {
        _httpClient = new HttpClient();
    }

    /// <summary>
    /// 查询Profile的订阅信息
    /// </summary>
    public async Task<SubscriptionInfo> QuerySubscriptionAsync(Profile profile)
    {
        if (profile.SubscriptionApi == null)
        {
            throw new InvalidOperationException("未配置订阅API");
        }

        var subscriptionInfo = profile.SubscriptionInfo ?? new SubscriptionInfo();
        subscriptionInfo.IsLoading = true;

        try
        {
            var token = GetTokenFromSettings(profile.Settings);
            var baseUrl = GetBaseUrlFromSettings(profile.Settings);

            // Cookie认证不需要token
            if (profile.SubscriptionApi.AuthType.ToLower() != "cookie")
            {
                if (string.IsNullOrEmpty(token))
                {
                    subscriptionInfo.HasError = true;
                    subscriptionInfo.ErrorMessage = "缺少认证 Token";
                    subscriptionInfo.IsLoading = false;
                    return subscriptionInfo;
                }
            }

            if (string.IsNullOrEmpty(baseUrl))
            {
                subscriptionInfo.HasError = true;
                subscriptionInfo.ErrorMessage = "缺少 BaseUrl";
                subscriptionInfo.IsLoading = false;
                return subscriptionInfo;
            }

            // 构建完整URL
            var fullUrl = CombineUrl(baseUrl, profile.SubscriptionApi.Endpoint);

            // 创建请求
            using var request = CreateHttpRequest(profile.SubscriptionApi, token, fullUrl);

            // 设置超时
            using var cts = new CancellationTokenSource(profile.SubscriptionApi.Timeout);

            // 发送请求
            var response = await _httpClient.SendAsync(request, cts.Token);
            response.EnsureSuccessStatusCode();

            // 解析响应
            var responseBody = await response.Content.ReadAsStringAsync();
            var jsonNode = JsonNode.Parse(responseBody);

            if (jsonNode == null)
            {
                throw new Exception("响应内容为空");
            }

            // 提取数据：data.items[0]
            var item = jsonNode["data"]?["items"]?[0];
            if (item == null)
            {
                throw new Exception("未找到订阅数据");
            }

            var subscriptionInfoNode = item["subscription_info"];
            if (subscriptionInfoNode == null)
            {
                throw new Exception("未找到订阅详情");
            }

            // 提取各个字段
            subscriptionInfo.DailyQuotaLimit = subscriptionInfoNode["daily_quota_limit"]?.GetValue<long>() ?? 0;
            subscriptionInfo.DailyQuotaUsed = item["daily_quota_used"]?.GetValue<long>() ?? 0;
            subscriptionInfo.WeeklyQuotaLimit = subscriptionInfoNode["weekly_quota_limit"]?.GetValue<long>() ?? 0;
            subscriptionInfo.WeeklyQuotaUsed = item["weekly_quota_used"]?.GetValue<long>() ?? 0;
            subscriptionInfo.TotalQuotaLimit = subscriptionInfoNode["total_quota_limit"]?.GetValue<long>() ?? 0;
            subscriptionInfo.TotalQuotaUsed = item["total_quota_used"]?.GetValue<long>() ?? 0;
            subscriptionInfo.ExpireTime = item["expire_time"]?.GetValue<long>() ?? 0;

            subscriptionInfo.HasError = false;
        }
        catch (OperationCanceledException)
        {
            subscriptionInfo.HasError = true;
            subscriptionInfo.ErrorMessage = "请求超时";
        }
        catch (HttpRequestException ex)
        {
            subscriptionInfo.HasError = true;
            subscriptionInfo.ErrorMessage = $"网络错误: {ex.Message}";
        }
        catch (Exception ex)
        {
            subscriptionInfo.HasError = true;
            subscriptionInfo.ErrorMessage = $"查询失败: {ex.Message}";
        }
        finally
        {
            subscriptionInfo.IsLoading = false;
        }

        return subscriptionInfo;
    }

    /// <summary>
    /// 从Settings中提取Token
    /// </summary>
    private string? GetTokenFromSettings(ClaudeSettings settings)
    {
        if (settings.ExtensionData?.TryGetValue("env", out var envElement) == true)
        {
            if (envElement.ValueKind == JsonValueKind.Object)
            {
                if (envElement.TryGetProperty("ANTHROPIC_AUTH_TOKEN", out var tokenElement))
                {
                    return tokenElement.GetString();
                }
            }
        }
        return null;
    }

    /// <summary>
    /// 从Settings中提取BaseUrl
    /// </summary>
    private string? GetBaseUrlFromSettings(ClaudeSettings settings)
    {
        if (settings.ExtensionData?.TryGetValue("env", out var envElement) == true)
        {
            if (envElement.ValueKind == JsonValueKind.Object)
            {
                if (envElement.TryGetProperty("ANTHROPIC_BASE_URL", out var baseUrlElement))
                {
                    return baseUrlElement.GetString();
                }
            }
        }
        return null;
    }

    /// <summary>
    /// 组合URL
    /// </summary>
    private string CombineUrl(string baseUrl, string endpoint)
    {
        baseUrl = baseUrl.TrimEnd('/');
        endpoint = endpoint.TrimStart('/');
        return $"{baseUrl}/{endpoint}";
    }

    /// <summary>
    /// 创建HTTP请求
    /// </summary>
    private HttpRequestMessage CreateHttpRequest(SubscriptionApiConfig apiConfig, string? token, string url)
    {
        var method = apiConfig.Method.ToUpper() == "POST" ? HttpMethod.Post : HttpMethod.Get;
        var request = new HttpRequestMessage(method, url);

        if (apiConfig.AuthType.ToLower() == "header")
        {
            // 使用Header认证：Authorization: Bearer {token}
            request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", token);
        }
        else if (apiConfig.AuthType.ToLower() == "body")
        {
            // 使用Body认证：{"token": "{token}"}
            var bodyContent = new { token };
            request.Content = JsonContent.Create(bodyContent);
        }
        else if (apiConfig.AuthType.ToLower() == "cookie")
        {
            // 使用Cookie认证
            if (!string.IsNullOrEmpty(apiConfig.SessionCookie))
            {
                request.Headers.Add("Cookie", apiConfig.SessionCookie);
            }

            // 添加额外的headers
            if (apiConfig.ExtraHeaders != null)
            {
                foreach (var header in apiConfig.ExtraHeaders)
                {
                    request.Headers.Add(header.Key, header.Value);
                }
            }
        }

        return request;
    }

    public void Dispose()
    {
        _httpClient?.Dispose();
    }
}

File: APISwitcher.macOS/APISwitcher/Services/SubscriptionService.swift
//
//  SubscriptionService.swift
//  APISwitcher
//
//  订阅信息查询服务
//

import Foundation

final class SubscriptionService: Sendable {
    private let session: URLSession

    init() {
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 30
        self.session = URLSession(configuration: config)
    }

    /// 查询订阅信息
    func querySubscription(config: Profile) async throws -> SubscriptionInfo {
        guard let subscriptionApi = config.subscriptionApi else {
            throw SubscriptionError.notConfigured
        }

        // 从 settings 中提取 token 和 baseUrl
        let token = getTokenFromSettings(config.settings)
        let baseUrl = getBaseUrlFromSettings(config.settings)

        // 检查认证信息（cookie 认证不需要 token）
        if subscriptionApi.authType.lowercased() != "cookie" {
            guard !token.isEmpty else {
                throw SubscriptionError.missingToken
            }
        }

        guard !baseUrl.isEmpty else {
            throw SubscriptionError.missingBaseUrl
        }

        // 构建完整 URL
        let fullUrl = try combineUrl(baseUrl: baseUrl, endpoint: subscriptionApi.endpoint)

        // 创建请求
        var request = createHttpRequest(
            apiConfig: subscriptionApi,
            token: token,
            url: fullUrl
        )

        // 设置超时
        request.timeoutInterval = TimeInterval(subscriptionApi.timeout) / 1000.0

        // 发送请求
        let (data, response) = try await session.data(for: request)

        // 检查响应状态
        guard let httpResponse = response as? HTTPURLResponse,
              (200...299).contains(httpResponse.statusCode) else {
            throw SubscriptionError.requestFailed
        }

        // 解析 JSON
        guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw SubscriptionError.invalidResponse
        }

        // 提取订阅数据：data.items[0]
        guard let dataDict = json["data"] as? [String: Any],
              let items = dataDict["items"] as? [[String: Any]],
              !items.isEmpty else {
            // API 调用成功，但没有订阅数据，表示用户未订阅
            return SubscriptionInfo(isUnsubscribed: true, isFirstLoad: false)
        }

        let item = items[0]

        guard let subscriptionInfoDict = item["subscription_info"] as? [String: Any] else {
            // 有订阅记录，但没有详细信息，也认为是未订阅
            return SubscriptionInfo(isUnsubscribed: true, isFirstLoad: false)
        }

        // 提取各个字段
        let dailyQuotaLimit = (subscriptionInfoDict["daily_quota_limit"] as? NSNumber)?.int64Value ?? 0
        let dailyQuotaUsed = (item["daily_quota_used"] as? NSNumber)?.int64Value ?? 0
        let weeklyQuotaLimit = (subscriptionInfoDict["weekly_quota_limit"] as? NSNumber)?.int64Value ?? 0
        let weeklyQuotaUsed = (item["weekly_quota_used"] as? NSNumber)?.int64Value ?? 0
        let totalQuotaLimit = (subscriptionInfoDict["total_quota_limit"] as? NSNumber)?.int64Value ?? 0
        let totalQuotaUsed = (item["total_quota_used"] as? NSNumber)?.int64Value ?? 0
        let expireTime = (item["expire_time"] as? NSNumber)?.int64Value ?? 0

        return SubscriptionInfo(
            dailyQuotaLimit: dailyQuotaLimit,
            dailyQuotaUsed: dailyQuotaUsed,
            weeklyQuotaLimit: weeklyQuotaLimit,
            weeklyQuotaUsed: weeklyQuotaUsed,
            totalQuotaLimit: totalQuotaLimit,
            totalQuotaUsed: totalQuotaUsed,
            expireTime: expireTime,
            isUnsubscribed: false,
            isFirstLoad: false
        )
    }

    /// 从 Settings 中提取 Token
    private func getTokenFromSettings(_ settings: ClaudeSettings) -> String {
        let dict = settings.toDictionary()
        if let env = dict["env"] as? [String: Any],
           let token = env["ANTHROPIC_AUTH_TOKEN"] as? String {
            return token
        }
        return ""
    }

    /// 从 Settings 中提取 BaseUrl
    private func getBaseUrlFromSettings(_ settings: ClaudeSettings) -> String {
        let dict = settings.toDictionary()
        if let env = dict["env"] as? [String: Any],
           let baseUrl = env["ANTHROPIC_BASE_URL"] as? String {
            return baseUrl
        }
        return ""
    }

    /// 组合 URL
    private func combineUrl(baseUrl: String, endpoint: String) throws -> URL {
        let cleanBase = baseUrl.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        let cleanEndpoint = endpoint.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        let urlString = "\(cleanBase)/\(cleanEndpoint)"

        guard let url = URL(string: urlString) else {
            throw SubscriptionError.invalidURL
        }
        return url
    }

    /// 创建 HTTP 请求
    private func createHttpRequest(apiConfig: SubscriptionApiConfig, token: String, url: URL) -> URLRequest {
        let method = apiConfig.method.uppercased()
        var request = URLRequest(url: url)
        request.httpMethod = method

        if apiConfig.authType.lowercased() == "header" {
            // 使用 Header 认证：Authorization: Bearer {token}
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        } else if apiConfig.authType.lowercased() == "body" {
            // 使用 Body 认证：{"token": "{token}"}
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
            let body = ["token": token]
            request.httpBody = try? JSONSerialization.data(withJSONObject: body)
        } else if apiConfig.authType.lowercased() == "cookie" {
            // 使用 Cookie 认证
            if let sessionCookie = apiConfig.sessionCookie, !sessionCookie.isEmpty {
                request.setValue(sessionCookie, forHTTPHeaderField: "Cookie")
            }

            // 添加额外的 headers
            if let extraHeaders = apiConfig.extraHeaders {
                for (key, value) in extraHeaders {
                    request.setValue(value, forHTTPHeaderField: key)
                }
            }
        }

        return request
    }
}

enum SubscriptionError: LocalizedError {
    case notConfigured
    case missingToken
    case missingBaseUrl
    case invalidURL
    case requestFailed
    case invalidResponse

    var errorDescription: String? {
        switch self {
        case .notConfigured:
            return "未配置订阅 API"
        case .missingToken:
            return "缺少认证 Token"
        case .missingBaseUrl:
            return "缺少 Base URL"
        case .invalidURL:
            return "无效的 API 地址"
        case .requestFailed:
            return "请求失败"
        case .invalidResponse:
            return "无效的响应数据"
        }
    }
}
